### ISSUE TITLE
manager.py chokes on special characters in dn

### ISSUE THREAD

**Issue by dirks** [dalibo/ldap2pg#114]
If the distinguished name of a user contains special characters, e.g. `CN=Markus Gräßer,OU=my_group,dc=example,dc=com` (name changed), execution of ldap2pg fails in apply_role_rules. The following error is thrown:

```
[ldap2pg.script   CRITICAL] Failed to process ('CN=Markus Gr\xc3\xa4\xc3\x9fer: 'ascii' codec can't decode byte 0xc3 in position 12: ordinal not in range(128) 
```

I worked around it by changing apply_role_rules
```python
def apply_role_rules(self, rules, entries):
    for rule in rules:
        for raw_entry in entries:
            entry = (
                raw_entry[0].decode('unicode_escape').encode('iso8859-1').decode('utf8'),
                raw_entry[1]
            )
            try:
                for role in self.process_ldap_entry(entry=entry, **rule):
                    yield role
            except ValueError as e:
                msg = "Failed to process %.32s: %s" % (entry, e,)
                raise UserError(msg)
```
The decoding line is a shameless copy from https://stackoverflow.com/a/6956914 (my python is limited).

This is on Centos 7.4.1708 with epel and [pgdg-96-centos](https://download.postgresql.org/pub/repos/yum/9.6/redhat/rhel-7.4-x86_64/) repository. ldap2pg was build according to https://ldap2pg.readthedocs.io/en/latest/install/ (the rpm in the pgdp-96 repo does not include all dependencies, but that is a bug report for the maintainer of the rpm).

**Comment by bersace** [comment]
Hi @dirks Thansk for pointing this ! I'm checking this.

**Comment by bersace** [comment]
@dirks, I'm missing something. Can you show me LDIF entry and `ldap2pg.yml` stanza to help me reproduce the bug ?

**Comment by dirks** [comment]
@bersace,  I changed a few values to make the entries anonymous in the example below. If you need an original entry I will send it via mail if that is fine with you. I did not notice earlier that the dn returned by ldapsearch is given as base64 string, if it contains Umlaute (ä, ß in this case).

```yaml
---
color: yes
verbose: yes
dry: yes
 
ldap:
    uri: ldap://ad1.example.com ldap://ad2.example.com
    binddn: cn=svc.mybindusers,ou=Functional Users,ou=Users,ou=EXAMPLE,dc=example,dc=com
    password: "redacted"
 
postgres:
    dsn: ""
    blacklist: [postgres, pg_*, testing]

sync_map:
    __all__:
    - ldap:
        base: ou=AAI,dc=example,dc=com
        filter: "(&(objectclass=User)(sAMAccountname=abc12de))"
        attributes: sAMAccountName
      role:
        name_attribute: sAMAccountName
        options: LOGIN
```

```
# extended LDIF
#
# LDAPv3
# base <ou=AAI,dc=example,dc=com> with scope subtree
# filter: (&(objectclass=User)(sAMAccountname=abc12de))
# requesting: sAMAccountName 
#

# Markus Gr\C3\A4\C3\9Fer, Admins, Employees, AAI, example.com
dn:: Q049TWFya3VzIEdyw6TDn2VyLE9VPUFkbWlucyxPVT1FbXBsb3llZXMsT1U9QUFJLERDPWV4YW1wbGUsREM9Y29tCg==
sAMAccountName: ab12cde

# search result
search: 5
result: 0 Success

# numResponses: 2
# numEntries: 1
```

**Comment by bersace** [comment]
@dirks yes, LDIF stores `UTF-8` as base64 with `::` as prefix like `dn:: …`. I'm checking this.

**Comment by bersace** [comment]
``` python
raw_entry[0].decode('unicode_escape').encode('iso8859-1').decode('utf8')
```

That looks ugly :/

Thanks for the base64. This looks easy to decode:

``` python
>>> my = base64.b64decode('Q049TWFya3VzIEdyw6TDn2VyLE9VPUFkbWlucyxPVT1FbXBsb3llZXMsT1U9QUFJLERDPWV4YW1wbGUsREM9Y29tCg==')
>>> my
b'CN=Markus Gr\xc3\xa4\xc3\x9fer,OU=Admins,OU=Employees,OU=AAI,DC=example,DC=com\n'
>>> my.decode('utf-8')
'CN=Markus Gräßer,OU=Admins,OU=Employees,OU=AAI,DC=example,DC=com\n'
```

(I just need one decode).

I still can't reproduce :/

Can you modify ldap2pg and reproduce with the following code in `apply_role_rules`:

``` python


    def apply_role_rules(self, rules, entries):
        for rule in rules:
            for entry in entries:
                try:
                    for role in self.process_ldap_entry(entry=entry, **rule):
                        yield role
                except ValueError as e:
                    # msg = "Failed to process %.32s: %s" % (entry, e,)
                    # raise UserError(msg)
                    raise
```

Danke schön :-)

**Comment by bersace** [comment]
@dirks That's strange to have a `\n` in `dn`.

**Comment by dirks** [comment]
The `\n` is an artifact from the name change. I used base64 from coreutils for the en/decoding and the input finished with a newline. Inside python I get the correct dn `Q049TWFya3VzIEdyw6TDn2VyLE9VPUFkbWlucyxPVT1FbXBsb3llZXMsT1U9QUFJLERDPWV4YW1wbGUsREM9Y29t`. I sent you the original LDIF via mail.

Here is the error thrown when using raw `raise`.

```
$ ldap2pg                                                                                                                                                                                                                                   [38/1916]
Starting ldap2pg 3.1.
Using /var/lib/pgsql/ldap2pg.yml.
[ldap2pg.ldap        DEBUG] Ignoring: [Errno 2] No such file or directory: u'/etc/ldap/ldap.conf'
[ldap2pg.ldap        DEBUG] Ignoring: [Errno 2] No such file or directory: u'/var/lib/pgsql/ldaprc'
[ldap2pg.ldap        DEBUG] Ignoring: [Errno 2] No such file or directory: u'/var/lib/pgsql/.ldaprc'
[ldap2pg.ldap        DEBUG] Ignoring: [Errno 2] No such file or directory: u'ldaprc'
[ldap2pg.ldap        DEBUG] Connecting to LDAP server ldap://ad1.example.com ldap://ad2.example.com
[ldap2pg.ldap        DEBUG] Trying simple bind.
[ldap2pg.script    WARNING] Running in dry mode. Postgres will be untouched.
[ldap2pg.manager      INFO] Inspecting Postgres...
[ldap2pg.psql        DEBUG] Connecting to Postgres dbname=postgres.
[ldap2pg.psql        DEBUG] Doing: SELECT datname FROM pg_catalog.pg_database
[ldap2pg.psql        DEBUG] WHERE datallowconn IS TRUE ORDER BY 1;
[ldap2pg.psql        DEBUG] Doing: SELECT
[ldap2pg.psql        DEBUG]     role.rolname, array_agg(members.rolname) AS members,
[ldap2pg.psql        DEBUG]     role.rolbypassrls, role.rolcanlogin, role.rolcreatedb, role.rolcreaterole, role.rolinherit, role.rolreplication, role.rolsuper
[ldap2pg.psql        DEBUG] FROM
[ldap2pg.psql        DEBUG]     pg_catalog.pg_roles AS role
[ldap2pg.psql        DEBUG] LEFT JOIN pg_catalog.pg_auth_members ON roleid = role.oid
[ldap2pg.psql        DEBUG] LEFT JOIN pg_catalog.pg_roles AS members ON members.oid = member
[ldap2pg.psql        DEBUG] GROUP BY role.rolname, role.rolbypassrls, role.rolcanlogin, role.rolcreatedb, role.rolcreaterole, role.rolinherit, role.rolreplication, role.rolsuper
[ldap2pg.psql        DEBUG] ORDER BY 1;
[ldap2pg.manager     DEBUG] Found role 'bc23defg' NOSUPERUSER NOBYPASSRLS NOCREATEDB INHERIT NOREPLICATION NOCREATEROLE LOGIN.
[ldap2pg.manager     DEBUG] Ignoring role pg_signal_backend. Matches u'pg_*'.
[ldap2pg.manager     DEBUG] Ignoring role postgres. Matches u'postgres'.
[ldap2pg.manager     DEBUG] Ignoring role testing. Matches u'testing'.
[ldap2pg.psql        DEBUG] Connecting to Postgres dbname=EIELAB.
[ldap2pg.psql        DEBUG] Doing: SELECT nspname FROM pg_catalog.pg_namespace
[ldap2pg.psql        DEBUG] WHERE nspname NOT LIKE 'pg_%' ORDER BY 1;
[ldap2pg.manager     DEBUG] Found schemas information_schema, public in EIELAB.
[ldap2pg.psql        DEBUG] Connecting to Postgres dbname=exampledb.
[ldap2pg.psql        DEBUG] Doing: SELECT nspname FROM pg_catalog.pg_namespace
[ldap2pg.psql        DEBUG] Doing: SELECT nspname FROM pg_catalog.pg_namespace
[ldap2pg.psql        DEBUG] WHERE nspname NOT LIKE 'pg_%' ORDER BY 1;
[ldap2pg.manager     DEBUG] Found schemas information_schema, public in exampledb.
[ldap2pg.psql        DEBUG] Using Postgres connection to dbname=postgres.
[ldap2pg.psql        DEBUG] Doing: SELECT nspname FROM pg_catalog.pg_namespace
[ldap2pg.psql        DEBUG] WHERE nspname NOT LIKE 'pg_%' ORDER BY 1;
[ldap2pg.manager     DEBUG] Found schemas information_schema, public in postgres.
[ldap2pg.psql        DEBUG] Connecting to Postgres dbname=template1.
[ldap2pg.psql        DEBUG] Doing: SELECT nspname FROM pg_catalog.pg_namespace
[ldap2pg.psql        DEBUG] WHERE nspname NOT LIKE 'pg_%' ORDER BY 1;
[ldap2pg.manager     DEBUG] Found schemas information_schema, public in template1.
[ldap2pg.manager     DEBUG] Postgres inspection done.
[ldap2pg.manager     DEBUG] Working on schema __all__.__any__.
[ldap2pg.manager      INFO] Querying LDAP ou=AAI,dc=example,dc=com...
[ldap2pg.ldap        DEBUG] Doing: ldapsearch -x -W -b ou=AAI,dc=example,dc=com -s sub '(&(objectclass=User)(sAMAccountname=abc12de))' sAMAccountName
[ldap2pg.manager     DEBUG] Got 1 entries from LDAP.
[ldap2pg.script      ERROR] Unhandled error:
[ldap2pg.script      ERROR] Traceback (most recent call last):
[ldap2pg.script      ERROR]   File "/opt/python-ldap2pg/lib/python2.7/site-packages/ldap2pg/script.py", line 75, in main
[ldap2pg.script      ERROR]     exit(wrapped_main(config))
[ldap2pg.script      ERROR]   File "/opt/python-ldap2pg/lib/python2.7/site-packages/ldap2pg/script.py", line 50, in wrapped_main
[ldap2pg.script      ERROR]     syncmap=config['sync_map'])
[ldap2pg.script      ERROR]   File "/opt/python-ldap2pg/lib/python2.7/site-packages/ldap2pg/manager.py", line 258, in inspect
[ldap2pg.script      ERROR]     for role in self.apply_role_rules(mapping['roles'], entries):
[ldap2pg.script      ERROR]   File "/opt/python-ldap2pg/lib/python2.7/site-packages/ldap2pg/manager.py", line 173, in apply_role_rules
[ldap2pg.script      ERROR]     for role in self.process_ldap_entry(entry=entry, **rule):
[ldap2pg.script      ERROR]   File "/opt/python-ldap2pg/lib/python2.7/site-packages/ldap2pg/manager.py", line 137, in process_ldap_entry
[ldap2pg.script      ERROR]     log_source = " from %s %s" % (entry[0], name_attribute)
[ldap2pg.script      ERROR] UnicodeDecodeError: 'ascii' codec can't decode byte 0xc3 in position 12: ordinal not in range(128)
[ldap2pg.script      ERROR] Please file an issue at https://github.com/dalibo/ldap2pg/issues with full log.
[ldap2pg.psql        DEBUG] Closing Postgres cursor to dbname=EIELAB.
[ldap2pg.psql        DEBUG] Closing Postgres connexion to dbname=EIELAB.
[ldap2pg.psql        DEBUG] Closing Postgres cursor to dbname=postgres.
[ldap2pg.psql        DEBUG] Closing Postgres connexion to dbname=postgres.
[ldap2pg.psql        DEBUG] Closing Postgres cursor to dbname=template1.
[ldap2pg.psql        DEBUG] Closing Postgres connexion to dbname=template1.
[ldap2pg.psql        DEBUG] Closing Postgres cursor to dbname=exampledb.
[ldap2pg.psql        DEBUG] Closing Postgres connexion to dbname=exampledb.
```

**Comment by dirks** [comment]
I noticed the error with users containing ä, ö, ü, ß, é in the CN part of dn.

**Comment by bersace** [comment]
Good news ! I reproduced it. It's related to python 2.7:

```
[ldap2pg.manager      INFO] Querying LDAP dc=ldap,dc=ldap2pg,dc=docker...
[ldap2pg.ldap        DEBUG] Doing: ldapsearch -x -W -b dc=ldap,dc=ldap2pg,dc=docker -s sub '(objectClass=uidObject)' uid
[ldap2pg.manager     DEBUG] Got 1 entries from LDAP.
[ldap2pg.script   CRITICAL] Failed to process ('cn=doroth\xc3\xa9e,ou=people,d: 'ascii' codec can't decode byte 0xc3 in position 9: ordinal not in range(128)
```

**Comment by bersace** [comment]
@dirks i just release 3.2 with the fix from #115 Can you test it ?

**Comment by dirks** [comment]
@bersace, yes the dry run works as expected. I continue deployment and testing this week. Thanks a lot. :)
